src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -0,0 +1,1042 @@
+	if (xgq->xgq_halted) {
+		XGQ_ERR(xgq, "xgq service is halted");
+		rval = -EIO;
+		goto done;
+	}
+
+	rval = xgq_produce(&xgq->xgq_queue, &slot_addr);
+	if (rval) {
+		XGQ_ERR(xgq, "error: xgq_produce failed: %d", rval);
+		goto done;
+	}
+
+	/* write head */
+	for (i = 0; i < sizeof(cmd->xgq_sq.xgq_sq_head) / 4; i++)
+		xgq_reg_write32(0, slot_addr + i * 4, *((u32 *)cmd + i));
+
+	/* write payload */
+	memcpy_toio((void __iomem *)(slot_addr + sizeof(cmd->xgq_sq.xgq_sq_head)),
+		&cmd->xgq_sq.xgq_sq_pkt, sizeof(cmd->xgq_sq.xgq_sq_pkt));
+
+	xgq_notify_peer_produced(&xgq->xgq_queue);
+
+	list_add_tail(&cmd->xgq_cmd_head, &xgq->xgq_submitted_cmds);
+
+done:
+	mutex_unlock(&xgq->xgq_lock);
+	return rval;
+}
+
+static void xgq_complete_cb(void *arg, struct xrt_com_queue_entry *ccmd)
+{
+	struct xocl_xgq_cmd *xgq_cmd = (struct xocl_xgq_cmd *)arg;
+
+	/* Note: we only care rcode for now */
+	xgq_cmd->xgq_cmd_rcode = ccmd->rcode;
+
+	complete(&xgq_cmd->xgq_cmd_complete);
+}
+
+/*
+ * Write buffer into shared memory and 
+ * return translate host based address to device based address.
+ * The 0 ~ XOCL_XGQ_RING_LEN is reserved for ring buffer.
+ * The XOCL_XGQ_DATA_OFFSET ~ end is for transferring shared data.
+ */
+static u64 memcpy_to_devices(struct xocl_xgq *xgq, const void *xclbin_data,
+	size_t xclbin_len)
+{
+	void __iomem *dst = xgq->xgq_ring_base + XOCL_XGQ_DATA_OFFSET;
+
+	memcpy_toio(dst, xclbin_data, xclbin_len);
+
+	/* This is the offset that device start reading data */
+	return XOCL_XGQ_DATA_OFFSET;
+}
+
+static inline int get_xgq_cid(struct xocl_xgq *xgq)
+{
+	int id;
+
+	mutex_lock(&xgq->xgq_lock);
+	id = idr_alloc_cyclic(&xocl_xgq_cid_idr, xgq, 0, 0, GFP_KERNEL);
+	mutex_unlock(&xgq->xgq_lock);
+
+	return id;
+}
+
+static inline void remove_xgq_cid(struct xocl_xgq *xgq, int id)
+{
+	mutex_lock(&xgq->xgq_lock);
+	idr_remove(&xocl_xgq_cid_idr, id);
+	mutex_unlock(&xgq->xgq_lock);
+}
+
+/*
+ * Utilize shared memory between host and device to transfer data.
+ */
+static ssize_t xgq_transfer_data(struct xocl_xgq *xgq, const void *u_xclbin,
+	u64 xclbin_len, xrt_xgq_pkt_type_t pkt_type, u32 timer)
+{
+	struct xocl_xgq_cmd *cmd = NULL;
+	struct xrt_sub_queue_entry *cmdp = NULL;
+	struct xgq_vmr_pkt *pkt;
+	ssize_t ret = 0;
+	int id;
+
+	cmd = kmalloc(sizeof(*cmd), GFP_KERNEL);
+	if (!cmd)
+		return ret;
+	memset(cmd, 0, sizeof(*cmd));
+
+	/*
+	 * xgq mangment driver will leverage XRT_CMD_OP_CONFIGURE cmd
+	 * and puts its own message inside the payload. Only RPU code
+	 * understand the format and decode it based on VERSION and TYPE.
+	 */
+	pkt = &(cmd->xgq_sq.xgq_sq_pkt);
+	pkt->head.version = 0;
+	pkt->head.type = pkt_type;
+	pkt->payload_xclbin.address = memcpy_to_devices(xgq, u_xclbin, xclbin_len);
+	pkt->payload_xclbin.size = xclbin_len;
+
+	cmd->xgq_cmd_cb = xgq_complete_cb;
+	cmd->xgq_cmd_arg = cmd;
+	cmd->xgq = xgq;
+
+	cmdp = &(cmd->xgq_sq.xgq_sq_head);
+	cmdp->opcode = XRT_CMD_OP_CONFIGURE;
+	cmdp->state = 1;
```

Comment:
Hmm...a comment might be helpful here indicating what a state of '1' represents.